### PR TITLE
Focus the correct window when closing a floating one

### LINKED ISSUE #545: Closing a floating window causes the underlaying layout to shift windows down

### PR DESCRIPTION
**Pull request by kynikos**
This patch fixes #545 for me, it fails the same tests as the develop branch, so I guess it's ok... ^^

### CODE REVIEW

**Comment by cjbarnes18** [comment]
Does this work when you set your layout to floating?

**Comment by kynikos** [comment]
Sorry for the delay but before answering I wanted to understand better how the whole focusing mechanism works. When the layout is set to floating, not only the patch that I originally pushed didn't fix the bug, but it was also making it worse because it could leave the manager without any focused windows.
Now, I think the only way to fix this bug for the floating layout would be to focus the previously focused window, but, as far as I understand it, Qtile doesn't keep the history of focused windows, i.e. the window "order" is static, so IMO at the current state this bug can only be fixed for non-floating layouts, which is what my new patch tries to do, without affecting the floating layout this time.
I'm not sure whether there's a better way to test if the current layout is floating though (the `is_layout_floating` method), do you have any better ideas?

File: libqtile/group.py [hunk 1]
```diff
@@ -201,12 +204,14 @@ def add(self, win, focus=True):
...
         self.windows.remove(win)
+        currentWindow = win.group.currentWindow
         win.group = None
         nextfocus = None
         if win.floating:
             nextfocus = self.floating_layout.remove(win)
             if nextfocus is None:
-                nextfocus = self.layout.focus_first()
+                nextfocus = self.layout.focus_first() \
+                    if self.is_layout_floating() else currentWindow
```
Comment (tych0) [inline libqtile/group.py]: This seems like a bit more of a general case: if the currentWindow is not win, we don't want to mess with the focus at all; should we just return after removing it above if that's the case?

**Comment by kynikos** [comment]
Uh I wonder why I didn't I think of this method before... The current patch, which should be pretty obvious, seems to fix the bug for me, including in the floating layout this time. I've also included your suggestion to return after removing in case currentWindow is not win.

**Comment by kynikos** [comment]
Ok, I've pushed a lot of amendments: please see how you like them.
Especially the last 3 introduce a history of focused windows: please see also the diff between the last amendmend and the current `develop` branch to see how little is actually changed overall.
If you like the focus history feature too, which can allow further expansion in the future, I think it would also make sense to merge `group._Group.focusHistory` with `group._Group.windows`
Once approved I can squash everything as requested.

**Comment by kynikos** [comment]
Sorry for the noise, I've noticed it's still buggy ^^'

**Comment by kynikos** [comment]
Ok, now everything seems to work all right, as I said also in the floating layout!
However I have a doubt at a8595b8d3f67a1b2fbb39181b31246d695456960: `currentWindow` can't be `None` anymore, so that assignment must be removed, but I don't even understand what it was doing there, so I wasn't sure how to replace it... In which cases is it possible that `focus` is called with a `win` that doesn't pass `if win:`? If it's only when there are no more windows in the group, then the following commits should be safe, but please triple check there.
Again, if you like it, I'll have to squash something and reword some commit messages before you merge I guess.

**Comment by tych0** [comment]
What problem does focushistory solve? I'm a little worried that it'll leak memory since windows are never really removed from it.

**Comment by kynikos** [comment]
As you may have noticed, the whole trick for fixing the bug is remembering the window that was focused before the floating one (that is going to be closed) last got the focus.
Until e5fa2cd3af571fe4c2ce0ca7e3ea0ab540df2e33 I did that by only remembering `previousWindow`, however, even though that worked, it didn't cover all the possible cases: it was in fact possible that `previousWindow` got closed without becoming `currentWindow`, for example a dialog that is closed programmatically by its application while another window is focused. This, even if it wouldn't break anything, wouldn't fix this bug in that particular case.
Introducing a list that keeps memory of the whole history in which the windows got the focus, `focusHistory`, I make sure that there will always be a `previousWindow`, regardless of the windows that are closed programmatically or not.
If `group._Group.remove` is always called when a window is closed, windows are always removed from `focusHistory`, see [group.py#L222](https://github.com/qtile/qtile/blob/c5eacdfc0fd3c6fae7d0039e2e1de81f584eaba7/libqtile/group.py#L222).
However I have pushed two new commits that should make this mechanism much clearer.
For the future, `focusHistory` could be used to introduce commands to navigate the windows based on the last time they received focus: I can imagine a `focus_previously_focused_window` command, or a new layout that makes use of this feature to order the windows according to `focusHistory`.

**Comment by kynikos** [comment]
I've rebased everything on top of a test that fails because of #545 and passes thanks to this patch.

**Comment by kynikos** [comment]
...and I've expanded the test to demonstrate the need for `focusHistory`: the test completely fails on the current develop branch; the specific focus history test fails at 619c564c559b222e0b74c89e7b8394647043d09e; the test passes at the last commit.

**Comment by tych0** [comment]
Thanks for the updates. I haven't had time to check this out yet, but I'll try to make some time this weekend.

**Comment by tych0** [comment]
Awesome. This looks great, thanks for taking the time to do it!

**Comment by kynikos** [comment]
Thank you for double-checking everything! Now however `_Group` stores two iterables of windows: I think it would make sense to merge `windows` and `focusHistory` into a unified object; I was thinking of a custom container-like object that preserves all the backward compatibility with the numerous references to `windows` also from the other modules. If you approve the idea, I can work on it (not on high priority) and open another PR when ready.

**Pull request closed without merging**